### PR TITLE
Add InfImputer

### LINKED ISSUE #358: Handling inf & -inf values
How to handle inf & -inf values in data ?
+-inf values are not treated as null by default in pandas, https://github.com/pandas-dev/pandas/issues/2858
Ideally inf value should be replaced by a large positive number and -inf should be replaced by negative large number, like we do in dataframe:

df.replace([np.inf, -np.inf],[np.finfo(np.float32).max,np.finfo(np.float32).min])

Choose np.float64 if that's the datatype.

Would be nice to do this using a step in Pipeline or some under the hood pre-processing step like we do fillna during reading.

### PR DESCRIPTION
**Pull request by milesgranger**
Will close #358

### CODE REVIEW

**Comment by choukha** [comment]
Works well with posinf but some issue with the neginf Imputation. I tried to pass a dataframe with neginf and posinf values. It passes through posinf, imputation, but getting error in neginf step. 
`TypeError: Must pass DataFrame or 2-d ndarray with boolean values only`
Or may be I'm doing some mistake here. Can you help/check ?

**Comment by milesgranger** [comment]
You're speedy at checking this, wasn't done yet, had to leave early today but I'll look into that aspect tomorrow! 👍

**Review by choukha — COMMENTED** [review]
Looks fine to me. :) 👍

**Review by epa095 — COMMENTED** [review]
Could you add a small test where it is part of a pipeline and de-serialized before use?

File: gordo_components/model/transformers/imputer.py [hunk 1]
```diff
@@ -26,19 +30,28 @@ def __init__(
...
-            where 'extremes' will use the min and max values for the current datatype.
-            such that 'inf' in a float32 dataset will have float32's largest value inserted.
+            choices: 'extremes', 'minmax'
+                -'extremes' will use the min and max values for the current datatype.
+                such that 'inf' in a float32 dataset will have float32's largest value inserted.
+                - 'minmax' will look at the min and max values in the feature where the -inf / inf
+                appears and fill with the max/min found in that feature.
+        power: float
+            Only applicable if ``strategy='minmax'``
+            Will compound the max/min value by this power. If the max value in a feature was 10
```
Comment (epa095) [inline gordo_components/model/transformers/imputer.py]: compound = multiply?

File: gordo_components/model/transformers/imputer.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    def get_params(self, deep=True):
+        return {
+            "inf_fill_value": self.inf_fill_value,
+            "neg_inf_fill_value": self.neg_inf_fill_value,
+            "strategy": self.strategy,
+            "power": self.power,
+        }
+
+    def fit(self, X: Union[pd.DataFrame, np.ndarray], y=None):
+        return self
```
Comment (epa095) [inline gordo_components/model/transformers/imputer.py:84]: here miles, miles!

File: gordo_components/model/transformers/imputer.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            choices: 'extremes', 'minmax'
+            -'extremes' will use the min and max values for the current datatype.
+            such that 'inf' in a float32 dataset will have float32's largest value inserted.
+            - 'minmax' will look at the min and max values in the feature where the -inf / inf
+            appears and fill with the max/min found in that feature.
+        power: float
+            Only applicable if ``strategy='minmax'``
+            Will compound the max/min value by this power. If the max value in a feature was 10
+            and ``power=2`` any inf value will be filled with 20. Likewise, if the min feature
+            was -10 any -inf will be filled with -20.
```
Comment (flikka) [inline gordo_components/model/transformers/imputer.py]: What if there is a -inf value, and the min feature is 2. Then -inf will be 4? I see from the code that this is not what will happen, maybe clarify the docs a bit there.

File: gordo_components/model/transformers/imputer.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            How to fill values, irrelevant if fill value is provided.
+            choices: 'extremes', 'minmax'
+            -'extremes' will use the min and max values for the current datatype.
+            such that 'inf' in a float32 dataset will have float32's largest value inserted.
+            - 'minmax' will look at the min and max values in the feature where the -inf / inf
+            appears and fill with the max/min found in that feature.
+        power: float
+            Only applicable if ``strategy='minmax'``
+            Will compound the max/min value by this power. If the max value in a feature was 10
+            and ``power=2`` any inf value will be filled with 20. Likewise, if the min feature
```
Comment (flikka) [inline gordo_components/model/transformers/imputer.py]: What if the max is negative? Will the positive inf then be power*a_negative_number?

File: gordo_components/model/transformers/imputer.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            for feature_idx, feature_max in enumerate(self._maxes):
+
+                # Skip features which don't have infs
+                if np.isposinf(X[:, feature_idx]).sum() == 0:
+                    continue
+
+                if max_allowable_value / self.power < feature_max:
+                    fill_value = max_allowable_value
+                else:
+                    fill_value = feature_max * self.power
```
Comment (flikka) [inline gordo_components/model/transformers/imputer.py]: Will this be correct if feature_max is negative?

**Review by flikka — CHANGES_REQUESTED** [review]

File: gordo_components/model/transformers/imputer.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            for feature_idx, feature_min in enumerate(self._mins):
+
+                # Skip features which don't have -infs
+                if np.isneginf(X[:, feature_idx]).sum() == 0:
+                    continue
+
+                if abs(feature_min) / self.power > abs(min_allowable_value):
+                    fill_value = min_allowable_value
+                else:
+                    fill_value = feature_min - (abs(feature_min) * self.power)
```
Comment (flikka) [inline gordo_components/model/transformers/imputer.py]: Wont this result in: if feature_min is -10, then the fill value will be -30?

File: gordo_components/model/transformers/imputer.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+class InfImputer(TransformerMixin):
+    def __init__(
+        self,
+        inf_fill_value=None,
+        neg_inf_fill_value=None,
+        strategy="minmax",
+        delta: float = 2.0,
+    ):
+        """
+        By default, fill -inf with lowest dtype value and inf with highest dtype value
```
Comment (flikka) [inline gordo_components/model/transformers/imputer.py]: Oops - some doc out of date here

File: gordo_components/model/transformers/imputer.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+        # Apply specific fill values if provided.
+        if self.inf_fill_value is not None:
+            X[np.isposinf(X)] = self.inf_fill_value
+        if self.neg_inf_fill_value is not None:
+            X[np.isneginf(X)] = self.neg_inf_fill_value
+
+        # May still be left over infs, if only one fill value was supplied for example
+        if self.strategy is not None:
+            return getattr(self, f"_fill_{self.strategy}")(X)
```
Comment (flikka) [inline gordo_components/model/transformers/imputer.py:99]: OOH, too advanced for me. What does this line do? Forget it, I think I see it. Many strange things one can do in Python...

File: tests/gordo_components/model/test_transformers.py [hunk 9]
```diff
@@ -33,3 +40,142 @@ def test_multiply_by_function_transformer(self):
...
+        # Get the maxes of those features in the imputed X
+        filled_maxes = X[:, features_with_pos_infs].max(axis=0)
+
+        # Get the previous maxes
+        previous_maxes = np.ma.masked_invalid(base_x[:, features_with_pos_infs]).max(
+            axis=0
+        )
+
+        # Now compare, that each new max is now double the size of the previous max these
+        # are where the inf values occur and would have be replaced with 2 * max of that feature
```
Comment (flikka) [inline tests/gordo_components/model/test_transformers.py]: Comment outdated

File: tests/gordo_components/model/test_transformers.py [hunk 10]
```diff
@@ -33,3 +40,142 @@ def test_multiply_by_function_transformer(self):
...
+        # Get the maxes of those features in the imputed X
+        filled_mins = X[:, features_with_neg_infs].min(axis=0)
+
+        # Get the previous maxes
+        previous_mins = np.ma.masked_invalid(base_x[:, features_with_neg_infs]).min(
+            axis=0
+        )
+
+        # Now compare, that each new max is now double the size of the previous max these
+        # are where the inf values occur and would have be replaced with 2 * max of that feature
```
Comment (flikka) [inline tests/gordo_components/model/test_transformers.py]: Comment is bad

**Review by flikka — APPROVED** [review]
Good stuff, except from the docs and comments!

File: gordo_components/model/transformers/imputer.py (hunk 8 above)
Comment (milesgranger) [reply gordo_components/model/transformers/imputer.py:99]: Ya, :man_shrugging: I think falling into the pattern of `if` `elif` `elif` ... is disgusting... IMO.

**Pull request merged**